File: src/main/java/top/chiloven/lukosbot2/commands/kemono/model/Embed.java
package top.chiloven.lukosbot2.commands.kemono.model;

import com.google.gson.JsonObject;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

/**
 * Represent an embed stuff in the post.
 *
 * @param url         the url of the embed
 * @param subject     the title
 * @param description the description
 */
public record Embed(
        String url,
        String subject,
        String description
) {
    public static @Nullable Embed fromSpecificPost(JsonObject obj) {
        try {
            return new Embed(
                    obj.get("url").getAsString(),
                    obj.get("subject").getAsString(),
                    obj.get("description").getAsString()
            );
        } catch (Exception e) {
            return null;
        }
    }

    public @NonNull String getString() {
        StringBuilder sb = new StringBuilder();

        sb.append("内嵌：").append(url).append("\n");
        if (!subject.equals("null")) {
            sb.append(subject).append("：\n");
        }
        if (!description.equals("null")) {
            sb.append(description).append("\n");
        }

        return sb.toString();
    }
}

File: src/main/java/top/chiloven/lukosbot2/commands/impl/kemono/schema/Embed.kt
package top.chiloven.lukosbot2.commands.impl.kemono.schema

import tools.jackson.databind.node.ObjectNode
import top.chiloven.lukosbot2.util.JsonUtils

data class Embed(
    val url: String = "",
    val subject: String? = null,
    val description: String? = null,
) {

    companion object {

        fun fromJsonObject(obj: ObjectNode): Embed = JsonUtils.snakeTreeToValue(obj, Embed::class.java)

    }

    fun getString(): String {
        return buildString {
            subject?.let { appendLine(it) }
            description?.let { appendLine(it) }
            appendLine(url)
        }.trim()
    }

}
